rstudio/shiny
# Add shiny::markdown() for inline Markdown

This is a stab at implementing https://github.com/rstudio/shiny/issues/2732 and incorporates implementation suggestions by @hadley.

The implementation is very small, but this does bring in the commonmark and glue packages as dependencies (but no transitive dependencies). 

- commonmark is widely used as a dependency of roxygen2 but is unlikely to be installed on user machines not geared toward package development.
- glue is part of the tidyverse, and so probably likelier to exist on user machines.
    - I prototyped a pure-R implementation of glue::trim(), but maintaining and testing it seems onerous (I include it at the bottom). It's also wildly inefficient compared to glue's C version.

## Open questions

* Should this be an addition to htmltools instead of to Shiny?
* Currently input strings are trimmed with `glue::trim()` but are not passed to `glue::glue()`, which would both trim them and support interpolation. If the strings were passed to `glue::glue()`, it's not obvious to me how additional arguments to `glue::glue()` could be passed, since additional arguments are currently passed as commonmark::markdown_html() arguments.

## Remaining dev tasks

- [x] Update NEWS

## `trim()` in R

```R
trim1 <- function(txt) {
  stopifnot(is.character(txt) && length(txt) == 1)
  # Remove leading and trailing whitespace
  txt <- gsub("^\\W*", "", txt)
  txt <- gsub("\\W*$", "", txt)
  # Determine and strip leading indentation
  lines <- strsplit(txt, "\n", fixed = TRUE)[[1]]
  rest <- tail(lines, -1)
  indents <- vapply(
    gregexpr("^\\W+", x),
    function (m) attr(m, "match.length"),
    numeric(1)
  )
  stripped <- gsub(paste0("^\\W{", min(indents), "}"), "", lines)
  paste0(stripped, collapse = "\n")
}
```

## QA Notes

1. Install Shiny from Github with `remotes::install_github("rstudio/shiny")`
2. Create a new single-file Shiny app in the RStudio IDE with `File > New File > Shiny Web App...`
3. At the top right of the generated `app.R` tap pane, press the "Run App" button. A window containing the Shiny demo app should pop up.
4. Close the demo app window and go back to the `app.R` tab.
5. Between lines 29 and 30, insert the following code:
```
markdown("# Some markdown
* This
* Is
* __MARKDOWN__
"),
```
6. Save `app.R` and press the "Run App" button again.
7. The demo app window should appear again. This time, you should see the markdown code you added rendered properly in the app, and it should look similar to the image below:

![Screen Shot 2020-04-07 at 8 39 18 PM](https://user-images.githubusercontent.com/26024/78742071-e0ba5680-790f-11ea-8bcb-da0e78e1e7fa.png)

8. If you make the following changes and save and run the app, the Markdown output should remain the same:

* Spaces before `#` in the line `# Some markdown` 
* Empty or blank lines before `# Some markdown`
* Empty or blank lines between `# Some markdown` and `* This`
* Empty or blank lines after the line `* __MARKDOWN__`
9. Generally, any markdown you can think to try should work and its rendered appearance should not surprise you.
